pgx-pg-sys/src/submodules/pg_try.rs

```diff
@@ -0,0 +1,94 @@
+use crate::errcodes::PgSqlErrorCode;
+use crate::panic::CaughtError;
+use std::collections::BTreeMap;
+use std::panic::{catch_unwind, UnwindSafe};
+
+pub struct PgTryBuilder<'a, R, F: FnOnce() -> R + UnwindSafe> {
+    func: F,
+    when: BTreeMap<PgSqlErrorCode, Box<dyn FnMut(CaughtError) -> R + 'a>>,
+    catch_others: Option<Box<dyn FnMut(CaughtError) -> R + 'a>>,
+    finally: Option<Box<dyn FnMut() + 'a>>,
+}
+
+impl<'a, R, F: FnOnce() -> R + UnwindSafe> PgTryBuilder<'a, R, F> {
+    #[must_use = "must call `PgTryBuilder::execute(self)` in order for it to run"]
+    pub fn new(func: F) -> Self {
+        Self { func, when: Default::default(), catch_others: None, finally: None }
+    }
+
+    #[must_use = "must call `PgTryBuilder::execute(self)` in order for it to run"]
+    pub fn catch_when(
+        mut self,
+        error: PgSqlErrorCode,
+        f: impl FnMut(CaughtError) -> R + 'a,
+    ) -> Self {
+        self.when.insert(error, Box::new(f));
+        self
+    }
+
+    #[must_use = "must call `PgTryBuilder::execute(self)` in order for it to run"]
+    pub fn catch_others(mut self, f: impl FnMut(CaughtError) -> R + 'a) -> Self {
+        self.catch_others = Some(Box::new(f));
+        self
+    }
+
+    #[must_use = "must call `PgTryBuilder::execute(self)` in order for it to run"]
+    pub fn finally(mut self, f: impl FnMut() + 'a) -> Self {
+        self.finally = Some(Box::new(f));
+        self
+    }
+
+    #[track_caller]
+    pub fn execute(mut self) -> R {
+        let result = catch_unwind(self.func);
+
+        let finally = || {
+            if let Some(mut finally) = self.finally {
+                // finally block runs
+                finally();
+            }
+        };
+
+        let result = match result {
+            Ok(result) => result,
+            Err(error) => {
+                let (sqlerrcode, panic) = match crate::panic::downcast_err(error) {
+                    panic @ CaughtError::RustPanic { .. } => (PgSqlErrorCode::RustPanic, panic),
+                    CaughtError::ErrorReport(ereport) => {
+                        let sqlerrcode = ereport.ereport.errcode;
+                        let panic = CaughtError::ErrorReport(ereport);
+                        (sqlerrcode, panic)
+                    }
+                    CaughtError::PostgresError(errdata) => {
+                        let sqlerrcode = errdata.sqlerrcode;
+                        let panic = CaughtError::PostgresError(errdata);
+                        (sqlerrcode, panic)
+                    }
+                };
+
+                // Postgres source docs says that a PG_TRY/PG_CATCH/PG_FINALLY block can't have
+                // both a CATCH and a FINALLY.
+                //
+                // We'll allow it, but if a catch handler (re)throws then the finally block will
```

Comment:
We should be sure to document this, because it's somewhat surprising that it's not run. Although, I guess I don't know what happens if you throw inside `finally` in like, Java/JavaScript/C# or whatever.

That said, I'm a fan of this behavior. If users *do* want a `finally` that always runs, they are free to do so by putting some type with a Drop impl on the stack (or via `scope_guard` for example).

Reply:
Indeed.  I'll be putting up some doc comments today that will address this.